src/core/datastore/persisters/key-value-store-persister.js

```diff
@@ -0,0 +1,49 @@
+import { Promise } from 'es6-promise';
+
+import { Client } from '../../client';
+import { NotFoundError, KinveyError } from '../../errors';
+
+import { KeyValuePersister } from './key-value-persister';
+
+export class KeyValueStorePersister extends KeyValuePersister {
```

Comment:
I'm confused by this class. What is it supposed to represent? What does it add to the `KeyValuePersister`? Other than the `_storeName`, I don't see any particular responsibility for this class.

Reply:
This is a persister, which can persist to key-value stores. IndexedDb is a key-value store. We are also using WebSQL and SQLite as key-value stores. What I mean by that is that we are writing and reading by key, the id of the entitiy, and the value being the entire entity (in SQL the entity is stringified).

These stores handle "byId" operations efficiently. However, we have querying which has to be done in memory (for `find()`, `remove()`, `count()`). So we need an API, which can, as efficiently as possible, read/write entire collections, as well as single entities from these collections.

The `KeyValuePersister` can read/write/delete by key and we use that for the "entire collection" part. `KeyValueStorePersister` extends on that, by adding methods for doing the same, on an entity level, thus utilizing the capabilities of a key/value store. It encapsulates working with the store entirely, abstracting away the database name (`_storeName` you mentioned) and working with the collections/entities in the most efficient way possible (all in a single transaction, etc).

Ontop of `KeyValueStorePersister` is the `KeyValueStoreOfflineRepository`, which extends upon `InmemoryOfflineRepository` and adds better performance for "byId" operations, utilizing the `KeyValueStorePersister`. It keeps the functionality from its parent, when it comes to querying operations, but overrides "byId" operations. Reusing the inmemory CRUD queue from parent, etc.